File: csharp/Prescribe.Core/Runtime/Defaults.cs
using Prescribe.Core.Semantics;
using Prescribe.Core.Util;

namespace Prescribe.Core.Runtime;

public static class Defaults
{
    public static RuntimeValue DefaultValue(TypeSymbol type)
    {
        switch (type.Kind)
        {
            case TypeKind.Integer:
                return RuntimeValues.Make(type, 0);
            case TypeKind.Real:
                return RuntimeValues.Make(type, 0.0);
            case TypeKind.Boolean:
                return RuntimeValues.Make(type, false);
            case TypeKind.Char:
                return RuntimeValues.Make(type, "\0");
            case TypeKind.String:
                return RuntimeValues.Make(type, "");
            case TypeKind.Date:
                return RuntimeValues.Make(type, new DateValue(1, 1, 1));
            case TypeKind.Array:
            {
                var arrayType = (ArrayType)type;
                var dims = arrayType.Bounds.Select(b => b.High - b.Low + 1).ToArray();
                object Create(int depth)
                {
                    var size = dims[depth];
                    var arr = new object[size];
                    if (depth == dims.Length - 1)
                    {
                        for (var i = 0; i < size; i += 1)
                        {
                            arr[i] = DefaultValue(arrayType.ElementType);
                        }
                    }
                    else
                    {
                        for (var i = 0; i < size; i += 1)
                        {
                            arr[i] = Create(depth + 1);
                        }
                    }
                    return arr;
                }
                return RuntimeValues.Make(type, Create(0));
            }
            case TypeKind.Record:
            {
                var recordType = (RecordType)type;
                var fields = new Dictionary<string, RuntimeValue>(StringComparer.Ordinal);
                foreach (var (name, fieldType) in recordType.Fields)
                {
                    fields[name] = DefaultValue(fieldType);
                }
                return RuntimeValues.Make(type, fields);
            }
            case TypeKind.Enum:
                return RuntimeValues.Make(type, 0);
            case TypeKind.Set:
                return RuntimeValues.Make(type, new HashSet<int>());
            case TypeKind.Pointer:
            case TypeKind.Class:
            case TypeKind.TextFile:
            case TypeKind.RandomFile:
                return RuntimeValues.Make(type, null);
            default:
                throw new InvalidOperationException("Unknown type default.");
        }
    }
}

File: legacy-typescript/src/runtime/defaults.ts
import { Type } from "../semantics/types.js";
import { makeValue, RuntimeValue } from "./values.js";
import { DateValue } from "../util/dates.js";

export function defaultValue(type: Type): RuntimeValue {
  switch (type.kind) {
    case "Integer":
      return makeValue(type, 0);
    case "Real":
      return makeValue(type, 0.0);
    case "Boolean":
      return makeValue(type, false);
    case "Char":
      return makeValue(type, "\u0000");
    case "String":
      return makeValue(type, "");
    case "Date":
      return makeValue(type, { year: 1, month: 1, day: 1 } as DateValue);
    case "Array": {
      const dims = type.bounds.map((b) => b.high - b.low + 1);
      const create = (depth: number): any => {
        const size = dims[depth];
        const arr = new Array(size);
        if (depth === dims.length - 1) {
          for (let i = 0; i < size; i += 1) {
            arr[i] = defaultValue(type.elementType);
          }
        } else {
          for (let i = 0; i < size; i += 1) {
            arr[i] = create(depth + 1);
          }
        }
        return arr;
      };
      return makeValue(type, create(0));
    }
    case "Record": {
      const fields = new Map<string, RuntimeValue>();
      for (const [name, fieldType] of type.fields.entries()) {
        fields.set(name, defaultValue(fieldType));
      }
      return makeValue(type, fields);
    }
    case "Enum":
      return makeValue(type, 0);
    case "Set":
      return makeValue(type, new Set<number>());
    case "Pointer":
      return makeValue(type, null);
    case "Class":
      return makeValue(type, null);
    case "TextFile":
      return makeValue(type, null);
    case "RandomFile":
      return makeValue(type, null);
  }
  throw new Error("Unknown type default.");
}
